subsys/net/lib/wifi_credentials/wifi_credentials.c

```diff
@@ -0,0 +1,538 @@
+		LOG_ERR("SSID has invalid format");
+		return -EINVAL;
+	}
+
+	k_mutex_lock(&wifi_credentials_mutex, K_FOREVER);
+	ssize_t idx = lookup_idx(ssid);
+
+	if (idx == -1) {
+		LOG_DBG("entry to delete was not found");
+		goto exit;
+	}
+
+	memset(configured_ssids[idx], 0, WIFI_SSID_MAX_LEN);
+
+	ret = wifi_credentials_delete_entry(idx);
+
+exit:
+	k_mutex_unlock(&wifi_credentials_mutex);
+	return ret;
+}
+
+int wifi_credentials_list_ssids(char *ssid_buf[], size_t ssid_buf_len)
+{
+	int ret = 0;
+
+	k_mutex_lock(&wifi_credentials_mutex, K_FOREVER);
+	if (ssid_buf_len < CONFIG_WIFI_CREDENTIALS_MAX_ENTRIES) {
+		ret = -EINVAL;
+		goto exit;
+	}
+	size_t j = 0;
+
+	for (size_t i = 0; i < CONFIG_WIFI_CREDENTIALS_MAX_ENTRIES; ++i) {
+		if (strlen(configured_ssids[i]) > 0) {
+			memcpy(ssid_buf[j], configured_ssids[i], WIFI_SSID_MAX_LEN);
+			j += 1;
+		}
+	}
+exit:
+	k_mutex_unlock(&wifi_credentials_mutex);
+	return ret;
+}
+
+#if defined(CONFIG_WIFI_CREDENTIALS_BACKEND_SETTINGS)
+#include <zephyr/settings/settings.h>
+
+BUILD_ASSERT(ENTRY_MAX_LEN <= SETTINGS_MAX_VAL_LEN);
+
+/**
+ * @brief WiFi credentials settings handler
+ * @note gets called when settings are loaded
+ *
+ * @param key credential index
+ * @param len length of entry
+ * @param read_cb read callback
+ * @param cb_arg argument for read callback
+ * @return 0 on success, otherwise a negative error code
+ */
+static int settings_set(const char *key, size_t len, settings_read_cb read_cb, void *cb_arg)
+{
+	/* key validation */
+	if (!key) {
+		return -EINVAL;
+	}
+
+	int idx = atoi(key);
+
+	if ((idx == 0 && strcmp(key, "0") != 0) ||
+	     idx >= CONFIG_WIFI_CREDENTIALS_MAX_ENTRIES) {
+		return -EINVAL;
+	}
+
+	if (len < sizeof(struct wifi_credentials_header)) {
+		return -EINVAL;
+	}
+
+	uint8_t buf[ENTRY_MAX_LEN];
+	size_t length_read = read_cb(cb_arg, buf, ARRAY_SIZE(buf));
+
+	/* value validation */
+	if (length_read < len) {
+		return -ENODATA;
+	}
+
+	load_entry(idx, (struct wifi_credentials_header *) buf);
+	return 0;
+}
+
+int wifi_credentials_serialize_entry(size_t idx, const void *buf, size_t buf_len)
+{
+	char settings_name_buf[16] = {0};
+
+	int ret = snprintk(settings_name_buf, ARRAY_SIZE(settings_name_buf), "wifi_cred/%d", idx);
+
+	if (ret < 0 || ret == ARRAY_SIZE(settings_name_buf)) {
+		LOG_ERR("settings key could not be generated");
+		return -EFAULT;
+	}
+
+	return settings_save_one(settings_name_buf, buf, buf_len);
+}
+
+int wifi_credentials_delete_entry(size_t idx)
+{
+	char settings_name_buf[16] = {0};
+
+	int ret = snprintk(settings_name_buf, ARRAY_SIZE(settings_name_buf), "wifi_cred/%d", idx);
+
+	if (ret < 0 || ret == ARRAY_SIZE(settings_name_buf)) {
+		LOG_ERR("settings key could not be generated");
+		return -EFAULT;
+	}
+
+	return settings_delete(settings_name_buf);
+}
+
+struct deserialize_helper_arg {
+	uint8_t *buf;
+	size_t buf_len;
+};
+
+static int deserialize_helper(const char *key, size_t len,
```

Comment:
I recommend renaming this to something more specific, such as

`zephyr_settings_backend_load_cb`

It also needs a comment explaining what it is and why you are using `settings_load_subtree_direct` instead of `settings_load_subtree`

Reply:
done